propolis/src/hw/chipset/i440fx.rs

```diff
@@ -178,10 +186,19 @@ impl I440Fx {
     }
 }
 impl Chipset for I440Fx {
-    fn pci_attach(&self, bdf: Bdf, dev: Arc<dyn pci::Endpoint>) {
-        assert!(bdf.bus.get() == 0);
-
-        self.pci_bus.attach(bdf, dev, Some(self.route_lintr(&bdf)));
+    fn pci_root_bus(&self) -> &pci::Bus {
+        &self.pci_bus
+    }
+    fn pci_router(&self) -> &Arc<pci::router::Router> {
+        &self.pci_router
+    }
+    fn pci_attach(
+        &self,
+        bus: &pci::Bus,
+        bdf: Bdf,
```

Comment:
Having both the `&Bus` and `Bdf` here is a little weird to me.  If we're attaching a device in the context of the chipset, it makes sense to me that it (the chipset) would at that point have knowledge of the bridge topology at that point, so only the `Bdf` would be required.

Alternatively, if we're in some future state where a bridge (and associated bus) is attached on the fly later, then devices might be attached directly to it (via a `Bus` ref).

What's your take on it, considering you've spend more time researching/thinking about the topic?

Reply:
I like the second approach, so much so that I implemented something like it in an earlier draft of this PR (but then excised it from this branch to limit this PR's scope). I pushed the commits to [gjcolombo/pci-bus-locations](https://github.com/oxidecomputer/propolis/tree/gjcolombo/pci-bus-locations) if you'd like to take a peek (needs cleanup before it can merge but should give you the gist).

I think a more general problem with bridges is that buses now have two names: the "logical" name/number the bus has in the bridge configuration in the TOML (or that you specify when you attach the bridge via the API, or what-have-you) and the number used to route PCI config traffic to the bus at runtime, which aren't guaranteed to be the same (and raises questions like "is the stored number in a `Bus` the former or the latter?"). So it makes a lot of sense to me to separate the questions of "how do I identify a specific bus?" from "given some bus, how do I address a specific device/function on that bus?", especially since you might not always use a `Bdf` in the former case (e.g. the server might instead refer to a bus by some logical name or entity ID because that's what's known to its client).